### DISCUSSION TITLE
How to perform vector searches on Milvus using Grpc?

### DISCUSSION THREAD

**Discussion by quachhengtony** [milvus-io/milvus#24280] [Q&A and General discussion] [answered]
I'm new to Grpc. I found the definition for search in https://github.com/milvus-io/milvus-proto/blob/2.2/proto/milvus.proto:
```
rpc Search(SearchRequest) returns (SearchResults) {}

message SearchRequest {
  option (common.privilege_ext_obj) = {
    object_type: Collection
    object_privilege: PrivilegeSearch
    object_name_index: 3
  };
  common.MsgBase base = 1; // must
  string db_name = 2;
  string collection_name = 3; // must
  repeated string partition_names = 4; // must
  string dsl = 5; // must
  // serialized `PlaceholderGroup`
  bytes placeholder_group = 6; // must
  common.DslType dsl_type = 7; // must
  repeated string output_fields = 8;
  repeated common.KeyValuePair search_params = 9; // must
  uint64 travel_timestamp = 10;
  uint64 guarantee_timestamp = 11; // guarantee_timestamp
  int64  nq = 12;
  bool not_return_all_meta = 13;
}
```
My question is where do I put the vectors?

**Comment by yhmo** [comment, marked as answer]
The vectors are converted to bytes and put into the "bytes placeholder_group = 6; // must"
In pymilvus:
https://github.com/milvus-io/pymilvus/blob/f6845fb8c51449633d9c68a559f64d83e01853ea/pymilvus/client/prepare.py#L462

In JAVA sdk:
https://github.com/milvus-io/milvus-sdk-java/blob/e3cca04112478dc3b276cfc22f97d28471821a1d/src/main/java/io/milvus/param/ParamUtils.java#L224

**Comment by quachhengtony** [comment]
How do I use the proto files for Grpc for .NET? Do I download and add the 3 files common.proto, schema.proto, and milvus.proto to my project's Protos folder, reference them in the csproj, and then run dotnet build? 
```
  <ItemGroup>
    <Protobuf Include="Protos\milvus.proto" GrpcServices="Client" />
    <Protobuf Include="Protos\common.proto" GrpcServices="Client" />
    <Protobuf Include="Protos\schema.proto" GrpcServices="Client" />
  </ItemGroup>
```

**Reply by yhmo** [reply]
There is already a c# sdk repo: https://github.com/milvus-io/milvus-sdk-csharp, it implements most of the APIs, but some new API of v2.2 are not ready. Some contributors are working on this sdk.

Milvus proto files are defined in this repo: https://github.com/milvus-io/milvus-proto

We require each sdk repo use the milvus proto as a git submdule. In the c# sdk repo, you can see the milvus-proto repos is included in gitmodules: https://github.com/milvus-io/milvus-sdk-csharp/blob/main/.gitmodules

The c# repo reference the proto files here: https://github.com/milvus-io/milvus-sdk-csharp/blob/02b99f009e52de9211648dc4ef3d45581bd0e5a1/src/IO.Milvus/IO.Milvus.csproj#L27

**Reply by quachhengtony** [reply]
Thanks for the help. I got it working.
